src/hammer-vlsi/technology/asap7/__init__.py

```diff
@@ -0,0 +1,156 @@
+                    mvt_cell = cell.copy(name=cell.name.replace('SL', vt), exclude_from_current=True, deep_copy=True)
+                # add mvt_cell to corresponding multi_lib
+                multi_lib['lib'].add(mvt_cell)
+
+        for vt, multi_lib in multi_libs.items():
+            # write multi_lib with 2.5e-10 precision
+            multi_lib['lib'].precision = 2.5e-10
+            multi_lib['lib'].write_gds(os.path.splitext(orig_gds)[0] + '_' + vt + '.gds')
+
+    def fix_sram_cdl_bug(self) -> None:
+        """
+        vendor's SRAM cdl use slvt cell, this patch will sed cells name in which, fix this bug.
+        """
+        self.logger.info("sed slvt to sram in asap7_75t_SRAM.cdl")
+        pattern0 = re.compile("SL")
+        pattern1 = re.compile("slvt")
+
+        with tempfile.NamedTemporaryFile(delete=False) as tf:
+            sram_cdl = os.path.join(self.extracted_tarballs_dir, "ASAP7_PDKandLIB.tar/ASAP7_PDKandLIB_v1p5/asap7libs_24.tar.bz2/asap7libs_24/cdl/lvs/asap7_75t_SRAM.cdl")
+            with open(sram_cdl, 'r') as f:
+                tf.write(pattern1.sub("sram", pattern0.sub("SRAM", f.read())).encode('utf-8'))
+            shutil.copystat(sram_cdl, tf.name)
+            shutil.copy(tf.name, sram_cdl)
+
+    def post_par_script(self, d: dict) -> None:
+        """
+        Scale the final GDS by a factor of 4
+        Called by the `run_par` method
+        """
+        self.logger.info("Scaling down place & routed GDS")
+
+        # load original_gds
+        par_gds_file = d.get("par.outputs.output_gds")
+        gds_lib = gdspy.GdsLibrary(infile=par_gds_file)
+        # Iterate through cells that aren't part of standard cell library and scale
+        for k,v in gds_lib.cell_dict.items():
+            if not 'ASAP7_75t' in k:
+                for poly in v.polygons:
+                    poly.scale(0.25)
+                for path in v.paths:
+                    path.scale(0.25)
+                    # gdspy bug: we also need to scale custom path extensions
+                    # wait gdspy/pull#101
+                    for i, end in enumerate(path.ends):
+                        if isinstance(end, tuple):
+                            path.ends[i] = tuple([e*0.25 for e in end])
+                for label in v.labels:
+                    # bug fix for some EDA tools that didn't set MAG field in gds file
+                    # maybe this is a excepted behavior in asap7 PDK
+                    # refer to gdspy/__init__.py: `kwargs["magnification"] = record[1][0]`
+                    label.magnification = 0.25
+                    label.translate(-label.position[0]*0.75, -label.position[1]*0.75)
+
+                for ref in v.references:
+                    ref.magnification = 0.25
+                    ref.translate(-ref.origin[0]*0.75, -ref.origin[1]*0.75)
+                    ref.magnification = 1
+
+        # Overwrite original GDS file & set precision to 2.5e-10
+        gds_lib.precision = 2.5e-10
```

Comment:
Don't we need to check that the precision was 1e-9 before doing this? The scaling was all done relatively so it should handle whatever precision the original gds had.

Reply:
Turns out that gdspy could not interpret the original precision in the P&R'd GDS. There might be a better way to do this, I'll investigate.

Reply:
gdspy can import the source units using `read_gds(units='import')` after all.